### PR TITLE
fix date picker controlled scenario

### PR DESCRIPTION
**Pull request by lorejoh12**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [x] Include a change request file using `$ yarn change`

#### Description of changes

adding fix for scenario where DatePicker is fully controlled and the internal selected date updates, but the formatted date does not. Also adding example to verify in the future.

Comments seem to suggest that updating the passed value prop to DatePicker was expected to call setSelectedDate somewhere, but value isn't a dependency of any useEffects and that function never gets called. Might have been left over from an earlier commit.

#### Focus areas to test

(optional)

### CODE REVIEW

File: packages/date-time/src/components/DatePicker/DatePicker.base.tsx [hunk 1]
```diff
@@ -487,7 +489,7 @@ export const DatePickerBase = React.forwardRef(
             {...calloutProps}
             className={css(classNames.callout, calloutProps && calloutProps.className)}
             // eslint-disable-next-line react/jsx-no-bind
-            onDismiss={calendarDismissed}
+            onDismiss={calloutDismissed}
```
Comment (lorejoh12) [inline packages/date-time/src/components/DatePicker/DatePicker.base.tsx:492]: unrelated bug fix that was annoying me- onDismiss accepts a function where the first parameter is a mouseEvent, but we were passing a function where the first parameter is an option Date. so the calendarDismissed function would get called with a MouseEvent instead of a Date and we'd get an error in the console

File: packages/date-time/src/components/DatePicker/DatePicker.base.tsx [hunk 2]
```diff
@@ -418,6 +416,10 @@ export const DatePickerBase = React.forwardRef(
       // don't need to focus the text box, if necessary the focusTrapZone will do it
     };
 
+    const calloutDismissed = (ev: React.MouseEvent<HTMLElement>): void => {
```
Comment (MLoughry) [inline packages/date-time/src/components/DatePicker/DatePicker.base.tsx:419]: What's the purpose for this added indirection?

Comment (lorejoh12) [reply packages/date-time/src/components/DatePicker/DatePicker.base.tsx:419]: looks like my earlier comment didn't get added in time, I'll copy/paste it here:

`unrelated bug fix that was annoying me- onDismiss accepts a function where the first parameter is a mouseEvent, but we were passing a function where the first parameter is an option Date. so the calendarDismissed function would get called with a MouseEvent instead of a Date and we'd get an error in the console`

I'm actually not sure why Typescript didn't catch this- perhaps because both parameters were optional, typescript says both being undefined means it's valid?

**Review by pompomon — APPROVED** [review]
